tests/lib/openstack_query/test_query_parser.py

```diff
@@ -0,0 +1,415 @@
+        Tests that run_parser functions expectedly - when giving group_by with no group_mappings
+        Should call build_unique_val_groups first to set group_mappings and then call run_group_by and return
+        """
+        self.instance._group_by = MockProperties.PROP_1
+        mock_obj_list = ["obj1", "obj2", "obj3"]
+        mock_build_unique_val_groups.return_value = "group-mapping-out"
+        mock_run_group_by.return_value = "group-out"
+
+        res = self.instance.run_parser(mock_obj_list)
+        self.assertEqual(res, "group-out")
+        mock_build_unique_val_groups.assert_called_once_with(
+            mock_obj_list, group_by_prop=MockProperties.PROP_1
+        )
+        mock_run_group_by.assert_called_once_with(mock_obj_list, "group-mapping-out")
+
+    @patch("openstack_query.query_parser.QueryParser._run_sort")
+    @patch("openstack_query.query_parser.QueryParser._run_group_by")
+    def test_run_parser_with_sort_and_group(self, mock_run_group_by, mock_run_sort):
+        """
+        Tests that run_parser functions expectedly - when giving both group_by and sort_by
+        Should call run_sort method and then run_group_by on that output, then return
+        """
+        self.instance._sort_by = (MockProperties.PROP_1, False)
+        self.instance._group_by = MockProperties.PROP_1
+        self.instance._group_mappings = {"group1": "some-mapping_func"}
+        mock_obj_list = ["obj1", "obj2", "obj3"]
+
+        mock_run_group_by.return_value = "group-out"
+        mock_run_sort.return_value = "sort-out"
+
+        res = self.instance.run_parser(mock_obj_list)
+        self.assertEqual(res, "group-out")
+        mock_run_sort.assert_called_once_with(mock_obj_list, self.instance._sort_by)
+        mock_run_group_by.assert_called_once_with(
+            "sort-out", self.instance._group_mappings
+        )
+
+    @parameterized.expand(
+        [
+            (
+                "string key ascending",
+                {"arg1": False},
+                [
+                    {"arg1": "a", "arg2": 2},
+                    {"arg1": "b", "arg2": 1},
+                    {"arg1": "c", "arg2": 4},
+                    {"arg1": "d", "arg2": 3},
+                ],
+            ),
+            (
+                "string key descending",
+                {"arg1": True},
+                [
+                    {"arg1": "d", "arg2": 3},
+                    {"arg1": "c", "arg2": 4},
+                    {"arg1": "b", "arg2": 1},
+                    {"arg1": "a", "arg2": 2},
+                ],
+            ),
+            (
+                "integer key ascending",
+                {"arg2": False},
+                [
+                    {"arg1": "b", "arg2": 1},
+                    {"arg1": "a", "arg2": 2},
+                    {"arg1": "d", "arg2": 3},
+                    {"arg1": "c", "arg2": 4},
+                ],
+            ),
+            (
+                "integer key descending",
+                {"arg2": True},
```

Comment:
Having seen this in use I think we should change the API as `True` doesn't mean anything

IMO the args should be something like:

```
class SortOrder(auto)
   ASC = auto()
   DEC = auto()
```

```
def sort(result, prop_order: List[str], order: SortOrder)
```

Then for the call is just
```
sort(results, ["name", "age"], SortOrder.ASC)
```

And it's much clearer than:
```
sort(results, {"name": true}
```

Reply:
no the API works like this:
```
sort(  ("name", True), ("age", True)  )
```
you're passing tuples not dicts 
We need to pass tuples because we want to allow users to sort by multiple keys in different orders

Currently because stackstorm is a pain I've defaulted to sorting every key by DESC 

The reason we use True and False is to emulate python `sort` which takes a `reverse` boolean flag - I can change it to `SortOrder` - that makes sense - I'll create an issue for it